src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsremandandsentencingapi/service/SentenceService.kt

```diff
@@ -182,4 +185,93 @@ class SentenceService(private val sentenceRepository: SentenceRepository, privat
       )
     }.toMutableSet()
   }
+
+  fun isTargetAlreadyInConsecutiveChain(
+    prisonerId: String,
+    appearanceUUID: UUID,
+    sourceSentenceUUID: UUID,
+    targetSentenceUUID: UUID,
+    sentencesOnAppearanceFromUI: List<SentenceDetailsForConsecValidation>,
+  ): Boolean {
+    if (sourceSentenceUUID == targetSentenceUUID) return true
+
+    sentencesOnAppearanceFromUI.forEach { log.info(it.toString()) }
+    // build upstream chains for the current appearance, starting from the sourceSentence
+    val upstreamChains = getUpstreamChains(sentencesOnAppearanceFromUI, sourceSentenceUUID)
+
+    // If target exists anywhere in those UI upstream chains then return true
+    if (upstreamChains.any { chain -> chain.any { it.sentenceUuid == targetSentenceUUID } }) {
+      log.info("targetSentenceUUID $targetSentenceUUID exists in upstream chain passed from UI")
+      return true
+    }
+
+    // source UUID's also to be checked against DB
+    val sourceUuids = upstreamChains
+      .flatten()
+      .map { it.sentenceUuid }
+      .distinct()
+
+    if (sourceUuids.isEmpty()) return false
+
+    // DB query to check if target appear downstream of the sources, any chains in the current appearance are omitted due tio UI check above
+    log.info("Checking target descendant outside current appearance")
```

Comment:
left this logging in - could be useful in the short term for testing and diagnostics